pkg/cloudprovider/providers/vsphere/vsphere_util.go

```diff
@@ -207,40 +207,47 @@ func (vs *VSphere) getSharedDatastoresInK8SCluster(ctx context.Context) ([]types
 	if err != nil {
 		return nil, err
 	}
-	dsMorefs := make(map[int][]types.ManagedObjectReference)
-	for i, vmMo := range vmMoList {
+	index := 0
+	var sharedDs []string
+	for _, vmMo := range vmMoList {
 		if !strings.HasPrefix(vmMo.Name, DummyVMPrefixName) {
-			dsMorefs[i], err = vs.getAllAccessibleDatastores(ctx, vmMo)
+			accessibleDatastores, err := vs.getAllAccessibleDatastores(ctx, vmMo)
 			if err != nil {
 				return nil, err
 			}
+			if index == 0 {
+				sharedDs = accessibleDatastores
+			} else {
+				sharedDs = intersect(sharedDs, accessibleDatastores)
+				if len(sharedDs) == 0 {
+					return nil, fmt.Errorf("No shared datastores found in the Kubernetes cluster")
+				}
+			}
+			index++
 		}
 	}
-	sharedDSMorefs := intersectMorefs(dsMorefs)
-	if len(sharedDSMorefs) == 0 {
-		return nil, fmt.Errorf("No shared datastores found in the Kubernetes cluster")
+	var sharedDSMorefs []types.ManagedObjectReference
+	for _, ds := range sharedDs {
+		sharedDSMorefs = append(sharedDSMorefs, types.ManagedObjectReference{
+			Value: ds,
+			Type:  "Datastore",
+		})
 	}
 	return sharedDSMorefs, nil
 }
 
-func intersectMorefs(args map[int][]types.ManagedObjectReference) []types.ManagedObjectReference {
-	arrLength := len(args)
-	tempMap := make(map[string]int)
-	tempArrayNew := make([]types.ManagedObjectReference, 0)
-	for _, arg := range args {
-		tempArr := arg
-		for idx := range tempArr {
-			if _, ok := tempMap[tempArr[idx].Value]; ok {
-				tempMap[tempArr[idx].Value]++
-				if tempMap[tempArr[idx].Value] == arrLength {
-					tempArrayNew = append(tempArrayNew, arg[idx])
-				}
-			} else {
-				tempMap[tempArr[idx].Value] = 1
+func intersect(list1 []string, list2 []string) []string {
+	var sharedList []string
+	for _, val1 := range list1 {
+		// Check if val1 is found in list2
+		for _, val2 := range list2 {
```

Comment:
This search will be expensive if the number of entries are a lot but may not be so inefficient if the number of entries are less. I was expecting you to use "map" since "set" data structure is not there in go std library. But the logic seems correct, see if you can use map instead.